internal/kubernetes/client.go

```diff
@@ -0,0 +1,126 @@
+package kubernetes
+
+import (
+	"path/filepath"
+	"regexp"
+	"strings"
+	"time"
+
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/cli-runtime/pkg/resource"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/kubectl/pkg/util"
+)
+
+const (
+	// Default cache directory.
+	cacheDir       = "/var/kube/cache"
+	defaultTimeout = 180 * time.Second
+)
+
+var (
+	ttl = 10 * time.Minute
+)
+
+type client struct {
+	c      dynamic.Interface
+	config *rest.Config
+	mapper *restmapper.DeferredDiscoveryRESTMapper
+	log    *zap.SugaredLogger
+}
+
+// apply will do create/patch of manifest
+func (c *client) apply(u *unstructured.Unstructured) error {
+	gvk := u.GroupVersionKind()
+	restMapping, err := c.mapper.RESTMapping(gvk.GroupKind(), gvk.Version)
+	if err != nil {
+		return err
+	}
+
+	gv := gvk.GroupVersion()
+	c.config.GroupVersion = &gv
+
+	restClient, err := newRestClient(*c.config, gv)
+	if err != nil {
+		return err
+	}
+
+	helper := resource.NewHelper(restClient, restMapping)
+
+	resourceInfo := &resource.Info{
+		Client:          restClient,
+		Mapping:         restMapping,
+		Namespace:       u.GetNamespace(),
+		Name:            u.GetName(),
+		Object:          u,
+		ResourceVersion: restMapping.Resource.Version,
+	}
+
+	patcher := newPatcher(resourceInfo, helper)
+
+	// Retrieves the modified configuration of the object, then embeds the result as an annotation in the modified configuration
+	modified, err := util.GetModifiedConfiguration(resourceInfo.Object, true, unstructured.UnstructuredJSONScheme)
+	if err != nil {
+		return err
+	}
+
+	// retrieves the object from the Namespace and Name fields to check if it already exists.
+	if err := resourceInfo.Get(); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+
+		// If object doesn't exist then create the resource in the given namespace.
+		obj, err := helper.Create(resourceInfo.Namespace, true, resourceInfo.Object)
+		if err != nil {
+			return err
+		}
+		if err := resourceInfo.Refresh(obj, true); err != nil {
+			c.log.Warn(err)
+		}
+	}
+
+	// Patch an OpenAPI resource
+	_, patchedObj, err := patcher.Patch(resourceInfo.Object, modified, resourceInfo.Source, resourceInfo.Namespace,
```

Comment:
need to run the patch in case of resource is not available and it's just created because patch will update the last applied annotation, like below:
```bash
   annotations:
    deployment.kubernetes.io/revision: "1"
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"annotations":{},"labels":{"app":"nginx","test":"nginx"},"name":"nginx","namespace":"numaplane-system"},"spec":{"replicas":1,"selector":{"matchLabels":{"app":"nginx"}},"template":{"metadata":{"labels":{"app":"nginx"}},"spec":{"containers":[{"image":"nginx","imagePullPolicy":"Always","name":"nginx"}]}}}}
```